pkg/meta/base.go

```diff
@@ -2808,3 +2810,98 @@ func inGroup(ctx Context, gid uint32) bool {
 	}
 	return false
 }
+
+type EntryScanner interface {
+	GetData(int) ([]*Entry, syscall.Errno)
+	Valid() bool
+	Close()
+}
+
+type mockScanner struct{}
+
+func (b *mockScanner) GetData(int) ([]*Entry, syscall.Errno) {
+	return nil, 0
+}
+
+func (b *mockScanner) Close() {
+}
+
+func (b *mockScanner) Valid() bool {
+	return false
+}
+
+type MockMetaEntryScannerGen struct{}
+
+func (b *MockMetaEntryScannerGen) NewMetaEntryScanner(inode Ino, plus bool) EntryScanner {
+	return &mockScanner{}
+}
+
+type baseEntrySanner struct {
+	c              EntryScanner
+	specialEntries []*Entry
+}
+
+func (sc *baseEntrySanner) Valid() bool {
+	return sc.c.Valid()
+}
+func (sc *baseEntrySanner) Close() {
+	sc.c.Close()
+}
+func (sc *baseEntrySanner) GetData(start int) ([]*Entry, syscall.Errno) {
+	if !sc.Valid() {
+		return nil, 0
+	}
+	res := make([]*Entry, 0)
+	if len(sc.specialEntries) > start {
+		res = append(res, sc.specialEntries[start:]...)
+		start = len(sc.specialEntries)
+	}
+	if entries, err := sc.c.GetData(start - len(sc.specialEntries)); err != 0 {
+		return make([]*Entry, 0), err
+	} else {
+		res = append(res, entries...)
+	}
+	return res, 0
+}
+
+func (m *baseMeta) NewBaseEntryScanner(ctx Context, inode Ino, plus uint8, InternalEntries []*Entry) (mc EntryScanner, rerr syscall.Errno) {
```

Comment:
1. maybe no need to nest a Scanner implementation.
just move kvEntryScanner.GetData to baseEntryScanner.getData

Reply:
Thank you for your suggestion. Based on my current understanding of juicefs code structure, we cannot operate directly on the engine at the **baseMeta** level. juicefs‘s architecture requires that it only call the functions specified in the engine interface, such as `doReaddir` and `doWrite`. Direct calls to TKV functions like `scanValues` or `gets` at this level are not feasible.

To address this, my current approach is to use a baseEntry to handle the interface boundaries.